util/reflect_test.go

```diff
@@ -856,6 +911,103 @@ func TestForEachField(t *testing.T) {
+			wantOut:      ``,
+		},
+		{
+			desc:         "nil fields",
+			parentStruct: &PointerOnlyBasicStruct{},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut:      ``,
+		},
+		{
+			desc:         "invalid path specification for field",
+			parentStruct: &PathErrorStruct{Field: String("value")},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut:      ``,
+			wantErr:      "field Field did not specify a path",
+		},
+		{
+			desc:         "struct without schema",
+			parentStruct: &basicStruct1,
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut:      `int32 : 42, string : "forty two", int32ptr : 4242, stringptr : "forty two ptr", `,
+		},
+		{
+			desc:         "struct of struct with no schema",
+			parentStruct: &StructOfStructs{BasicStructField: basicStruct1, BasicStructPtrField: &basicStruct2},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut: `int32 : 42, string : "forty two", int32ptr : 4242, stringptr : "forty two ptr", ` +
+				`int32 : 43, string : "forty three", int32ptr : 4343, stringptr : "forty three ptr", `,
+		},
+		{
+			desc:         "struct of slice of structs with no schema",
+			parentStruct: &StructOfSliceOfStructs{BasicStructSliceField: []BasicStruct{basicStruct1}, BasicStructPtrSliceField: []*BasicStruct{&basicStruct2}},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut: `int32 : 42, string : "forty two", int32ptr : 4242, stringptr : "forty two ptr", ` +
+				`int32 : 43, string : "forty three", int32ptr : 4343, stringptr : "forty three ptr", `,
+		},
+		{
+			desc:         "struct with string leaf-list",
+			parentStruct: &BasicSliceStruct{StringSlice: []string{"one", "two"}},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut:      ``,
+		},
+		{
+			desc:         "map keys with no struct schema",
+			in:           nil,
+			parentStruct: &StructOfMapOfStructs{BasicStructMapField: map[string]BasicStruct{"basicStruct1": basicStruct1}, BasicStructPtrMapField: map[string]*BasicStruct{"basicStruct2": &basicStruct2}},
+			iterFunc:     printMapKeysSchemaAnnotationFunc,
+			wantOut: `basicStruct1 (string)/basic-struct : 
+{Int32Field:     42,
+ StringField:    "forty two",
+ Int32PtrField:  4242,
+ StringPtrField: "forty two ptr"} (string)
+, basicStruct2 (string)/basic-struct : 
+{Int32Field:     43,
+ StringField:    "forty three",
+ Int32PtrField:  4343,
+ StringPtrField: "forty three ptr"} (string)
+, `,
+		},
+	}
+
+	for _, tt := range tests {
+		outStr := ""
+		var errs Errors
+		errs = ForEachDataField(tt.parentStruct, tt.in, &outStr, tt.iterFunc)
+		if got, want := errs.String(), tt.wantErr; got != want {
+			t.Errorf("%s: ForEachDataField(%v, %#v, ...): did not get expected error, got: %s, want: %s", tt.desc, tt.parentStruct, tt.in, got, want)
+		}
+		if errs == nil {
```

Comment:
If Errors is a list, we should use len() to check that it is empty, since a non-nil list can still be empty.

if len(errs) == 0 { ... }

On the other hand, in Go we usually indent the error-flow. So let's consider the opposite logic:

if len(errs) > 0 {
  continue
}

if got, want := ...

Reply:
Done, thanks.